leshan-core/src/main/java/org/eclipse/leshan/core/util/X509CertUtil.java

```diff
@@ -0,0 +1,310 @@
+     * Notes: Only understands X500Principal.
+     *
+     * @param principal Source principal.
+     * @param field     Field key as defined in RFC 2253.
+     * @return null or value as string.
+     */
+    public static String getPrincipalField(Principal principal, String field) {
+        if (principal instanceof X500Principal) {
+            X500Principal x500Principal = (X500Principal) principal;
+
+            // Extra practical OID that Java implementation hides for "not being standard" even thou
+            // list in RFC is only example OIDs list (see JDK class AVAKeyword). These extra OID's are
+            // actually defined in Sun packages but we should not access them as they are not standard
+            // Java classes.
+            HashMap<String, String> extraOids = new HashMap<String, String>();
+            extraOids.put("2.5.4.5", "SERIALNUMBER");
+
+            String dn = x500Principal.getName(X500Principal.RFC2253, extraOids);
+
+            Map<String, String> fields = parseRfc2253Name(dn);
+
+            if (fields.containsKey(field)) {
+                return fields.get(field);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * DNS name matcher
+     * <p>
+     * Supports matching with wildcard DNS names.
+     *
+     * @param matcher Matcher pattern
+     * @param target  Target DNS name to check
+     * @return true if matches, false otherwise
+     */
+    private static boolean dnsNameMatch(String matcher, String target) {
+        if (matcher.startsWith("*.")) {
+            // Simple filtering out
+            if (!target.endsWith(matcher.substring(1)))
+                return false;
+
+            // Wildcards only work in one sub level so no extra dots
+            String host = target.substring(0, target.length() - (matcher.length() - 1));
+            return host.indexOf('.') == -1;
+        }
+        return matcher.equals(target);
+    }
+
+    /**
+     * Match DNS name against X.509 Certificate's Subjects
+     *
+     * @param certificate Target X.509 certificate
+     * @param dnsName     DNS name to match
+     * @return True if match, false otherwise
+     */
+    public static boolean matchSubjectDnsName(X509Certificate certificate, String dnsName) {
+        try {
+            // First one need to check SANs if they are present
+
+            final Collection<List<?>> sans = certificate.getSubjectAlternativeNames();
+
+            if (sans != null) {
+                for (final List<?> san : sans) {
+                    int generalName = (Integer) san.get(0);
+                    if (generalName == GeneralName.DNS_NAME.value) {
+                        String value = (String) san.get(1);
+                        if (dnsNameMatch(value, dnsName)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            // Then one can fallback to old ways at looking in Subject DN
```

Comment:
Should we fallback :
-  only if san/dnsname does not exist 
OR 
- even if it exists and it does'nt match ?

Reply:
https://tools.ietf.org/html/rfc6125 seems to address the issue how it should be interpreted.

> Does your technology need to use DNS domain names in the Common
> Name of certificates for the sake of backward compatibility?  If
> so, consider recommending support for the CN-ID identifier type as
> a fallback."

Reply:
Let me try to reformulate because maybe I was not clear.

Imagine you want to check "server.mydomain.com"  domain.
And you have a certificate with SAN=dns:**another**.mydomain.com and CN=server.mydomain.com
Should we consider that as a match or not ?

Reply:
I just raise this question. We don't need to answer to this question quickly.
I will merge the PR in master.

Reply:
Yes, I understood your question.

SAN is list any-match-here is good.

In purest form answer would be no match for server.mydomain.com -- as if SAN is present it should be used as single truth.

However people is still trying to learn how the SAN should be configured so starting with SAN -> CN fallback might be the good start to make sure that no existing installations gets broken.

Making it more strict is probably a good idea and I believe some software are like that already. Making it more strict right away might be also a good option to force people to make things correctly.

When looking form openssl:
https://www.openssl.org/docs/man1.1.1/man3/X509_check_host.html

It says that any CN or SAN is good enough in generic text but then again has control flags that can flex or strict the operation. And even says multiple default operations that are a bit conflicting ;).

Reply:
Maybe we should start with strict mode and if some user complain we will try to adapt it (make it configurable)?

Reply:
Sure. I will make a PR for that.

Reply:
And the PR: https://github.com/eclipse/leshan/pull/910